game-core/src/main/java/games/strategy/engine/data/unit/ability/UnitAbilityFactory.java

```diff
@@ -0,0 +1,456 @@
+package games.strategy.engine.data.unit.ability;
+
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.FIRST_STRIKE_UNITS;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.NAVAL_BOMBARD;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.UNITS;
+
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.PlayerList;
+import games.strategy.engine.data.UnitType;
+import games.strategy.engine.data.UnitTypeList;
+import games.strategy.engine.data.battle.phase.BattlePhase;
+import games.strategy.engine.data.battle.phase.BattlePhaseList;
+import games.strategy.engine.data.properties.GameProperties;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import lombok.experimental.UtilityClass;
+
+/**
+ * Auto-generates unit abilities from deprecated unit options
+ *
+ * <p>If a map defines the unit abilities, then this should not be used.
+ */
+@UtilityClass
+public class UnitAbilityFactory {
+
+  private static final String WILL_NOT_FIRE_AA_ABILITY_PREFIX = "willNotFireAa";
+
+  /**
+   * Tracks the typeAa and their targetsAa
+   *
+   * <p>The typeAa and their targetsAa is copied on every single unitAttachment. This means that a
+   * map maker might have different targetsAa in the same typeAa. But that is a typo since the
+   * engine only uses the first targetsAa that it finds. This map keeps track of the typeAa and
+   * targetsAa separately from the unitAttachment so that there is a central place to find them.
+   */
+  private static final Map<String, Collection<UnitType>> aaTargets = new HashMap<>();
+
+  /**
+   * Tracks what unit types a specific unit type can not target without an ally destroyer present
+   *
+   * <p>The deprecated property canNotBeTargetedBy was used to determine if a unit could be targeted
+   * depending on the presence of an isDestroyer. This means that canNotBeTargetedBy is structured
+   * as targeted unit -> firing units. But unitAbilities are structured as firing unit -> targeted
+   * units.
+   *
+   * <p>This map is used to invert that data set at the beginning so that it can be easily looked up
+   * later
+   */
+  private static final Map<UnitType, Collection<UnitType>> unitCanNotTargetWithoutDestroyer =
+      new HashMap<>();
+
+  /**
+   * Create the default unit abilities at the beginning of the battle step
+   *
+   * <p>This needs to be called once at the beginning of each battle step. All the battles in the
```

Comment:
Noting the use of static state here. I'm very concerned this could get us into trouble. Any thoughts to make this stateful and by-battle?

Reply:
When you say "by-battle", you want this logic to be run at the beginning of each battle?  It isn't needed.  This is static information at the beginning of the entire battle phase and re-doing it for each battle would be unnecessary.

Reply:
Didn't notice you responded. My concerns:
1. We're relying on establishing static state, which depends on call ordering,  it is inherently brittle. 
2. Won't work when we have multiple games per process

I'm not sure what the best way to restructure would be to avoid the static state. Perhaps a transient game data member variable could do the trick. Another thought, is the battle phase represented as an object in any way? Perhaps that could have the logic to lazy compute the needed unit ability information.

Reply:
@DanVanAtta, no problem with the late reply.

The `BattlePhaseList` will be a member variable of something (probably `GameData`) and it will actually store the state.  This class is just converting the existing attributes into unit abilities.  So it really isn't state.  But there do happen to be two static private members I added to simplify some of the logic and those will cause problems with multiple games per process.  So I'll need to pass that data around instead of creating a private static member.

Also, if/when map makers can create custom unit abilities, then this class will not be used.  The xml parser will need to error out if it finds both custom unit abilities and the deprecated properties.  It could possibly use this class to give helpful hints on how to convert the deprecated properties.

Reply:
Will this comment change perhaps?

> This needs to be called once at the beginning of each battle step

I have also been wondering if even in an intermediate form if that would work. What if for example a battle is persisted at an unexpected time, would we still be able to guarantee all values are instantiated properly?

Reply:
No, I don't see that comment changing unless the existing code flow changes even more drastically.  The `BattlePhaseList` will most likely be stored as part of the save game so when the battle starts up again, it will restore it.  And if the code changes enough so that `BattlePhaseList` isn't part of the save game, when the save game is restored, it will be initialized when the Battle Step code is restored.

Reply:
I've removed the static members and now pass the needed state around - https://github.com/triplea-game/triplea/pull/8646/commits/4773376d77361c893b0d5d2490104bcd852c4921

Reply:
> Will this comment change perhaps?
> 
> > This needs to be called once at the beginning of each battle step

Also, as long as we have to support old map xmls, this code will need to exist.  If there aren't custom unit abilities, then this code has to be called so that it can convert the old properties into the unit abilities.  It is just adding these unit abilities to the `BattlePhaseList`. If custom unit abilities exist, then the `BattlePhaseList` will already be set up correctly.  Saving and loading of the `BattlePhaseList` is independent of this code since this is just conversion code.

Reply:
> Also, as long as we have to support old map xmls, this code will need to exist. 

FWIW, the design of the map-data project is such that it provides a unified data interface to the rest of the game. Any legacy XML attributes (or lack there-of), should be handled within that layer. This makes it transparent to the rest of the engine which XML attributes were present exactly.

Reply:
If this is conversion code, does it actually more properly belong in the parsing layer? There are ways to alias XML properties and/or straight up translate them from one to another. The intent is that map parsing can give a relatively idealized data interface and can "add-in" tags or translate tags from one property to another, or combine multiple tags to set the same data.

Reply:
I have thought about putting it in the parsing layer but decided not to because of the increase in complexity.  All of the properties that this code depends on can change with triggers.  And some of the properties are technology dependent so players can have different properties depending on what technologies they've researched.  I would have to write code that would run when a trigger happens or a technology is researched.  This code would have to rewrite the unit abilities depending on what changed.  That felt really complex compared to what I have now.

Reply:
Okay, fair enough. You're aware of that design element and it not working out is what it is. :+1: